ELF Loader Apple Compilation Issues

From the CI. Would be good for someone with a mac to fix.

```
2021-04-27T10:18:33.9872880Z FAILED: iree/hal/local/elf/CMakeFiles/iree_hal_local_elf_platform.dir/platform/apple.c.o 
2021-04-27T10:18:33.9987890Z /usr/bin/clang  -isystem /Users/runner/work/iree/iree/main_checkout -isystem . -O3 -DNDEBUG -isysroot /Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk -mmacosx-version-min=10.9 -fPIC -Wall -Wno-ambiguous-member-template -Wno-char-subscripts -Wno-deprecated-declarations -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end -Wno-ignored-optimization-argument -Wno-invalid-offsetof -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-pointer-sign -Wno-reserved-user-defined-literal -Wno-return-type-c-linkage -Wno-self-assign-overloaded -Wno-sign-compare -Wno-signed-unsigned-wchar -Wno-strict-overflow -Wno-trigraphs -Wno-unknown-pragmas -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-unused-const-variable -Wno-unused-function -Wno-unused-local-typedef -Wno-unused-private-field -Wno-user-defined-warnings -Wctad-maybe-unsupported -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wformat-security -Wgnu-redeclared-enum -Wimplicit-fallthrough -Winfinite-recursion -Wliteral-conversion -Wnon-virtual-dtor -Woverloaded-virtual -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety -Wthread-safety-beta -Wunused-comparison -Wvla -Wno-strict-prototypes -Wno-shadow-uncaptured-local -Wno-gnu-zero-variadic-macro-arguments -Wno-shadow-field-in-constructor -Wno-unreachable-code-return -Wno-missing-variable-declarations -Wno-gnu-label-as-value -Wno-unused-parameter -Wno-unused-variable -Wno-undef -fvisibility=hidden -std=gnu11 -MD -MT iree/hal/local/elf/CMakeFiles/iree_hal_local_elf_platform.dir/platform/apple.c.o -MF iree/hal/local/elf/CMakeFiles/iree_hal_local_elf_platform.dir/platform/apple.c.o.d -o iree/hal/local/elf/CMakeFiles/iree_hal_local_elf_platform.dir/platform/apple.c.o -c /Users/runner/work/iree/iree/main_checkout/iree/hal/local/elf/platform/apple.c
2021-04-27T10:18:34.0076820Z /Users/runner/work/iree/iree/main_checkout/iree/hal/local/elf/platform/apple.c:47:40: warning: 'pthread_jit_write_protect_np' is only available on macOS 11.0 or newer [-Wunguarded-availability-new]
2021-04-27T10:18:34.0175810Z void iree_memory_jit_context_begin() { pthread_jit_write_protect_np(0); }
2021-04-27T10:18:34.0275940Z                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-04-27T10:18:34.0379830Z /Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/pthread.h:561:6: note: 'pthread_jit_write_protect_np' has been marked as being introduced in macOS 11.0 here, but the deployment target is macOS 10.9.0
2021-04-27T10:18:34.0400530Z void pthread_jit_write_protect_np(int enabled);
2021-04-27T10:18:34.0501980Z      ^
2021-04-27T10:18:34.0605080Z /Users/runner/work/iree/iree/main_checkout/iree/hal/local/elf/platform/apple.c:47:40: note: enclose 'pthread_jit_write_protect_np' in a __builtin_available check to silence this warning
2021-04-27T10:18:34.0638190Z void iree_memory_jit_context_begin() { pthread_jit_write_protect_np(0); }
2021-04-27T10:18:34.0840180Z                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-04-27T10:18:34.0943810Z /Users/runner/work/iree/iree/main_checkout/iree/hal/local/elf/platform/apple.c:49:38: warning: 'pthread_jit_write_protect_np' is only available on macOS 11.0 or newer [-Wunguarded-availability-new]
2021-04-27T10:18:34.1043470Z void iree_memory_jit_context_end() { pthread_jit_write_protect_np(1); }
2021-04-27T10:18:34.1146000Z                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-04-27T10:18:34.1248600Z /Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/pthread.h:561:6: note: 'pthread_jit_write_protect_np' has been marked as being introduced in macOS 11.0 here, but the deployment target is macOS 10.9.0
2021-04-27T10:18:34.1348990Z void pthread_jit_write_protect_np(int enabled);
2021-04-27T10:18:34.1450750Z      ^
2021-04-27T10:18:34.1552990Z /Users/runner/work/iree/iree/main_checkout/iree/hal/local/elf/platform/apple.c:49:38: note: enclose 'pthread_jit_write_protect_np' in a __builtin_available check to silence this warning
2021-04-27T10:18:34.1652750Z void iree_memory_jit_context_end() { pthread_jit_write_protect_np(1); }
2021-04-27T10:18:34.1754530Z                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-04-27T10:18:34.1807540Z /Users/runner/work/iree/iree/main_checkout/iree/hal/local/elf/platform/apple.c:74:15: error: use of undeclared identifier 'IREE_MEMORY_VIEW_FLAG_EXECUTE'; did you mean 'IREE_MEMORY_VIEW_FLAG_MAY_EXECUTE'?
2021-04-27T10:18:34.1909800Z   if (flags & IREE_MEMORY_VIEW_FLAG_EXECUTE) {
2021-04-27T10:18:34.2009960Z               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-04-27T10:18:34.2111510Z               IREE_MEMORY_VIEW_FLAG_MAY_EXECUTE
2021-04-27T10:18:34.2213230Z /Users/runner/work/iree/iree/main_checkout/iree/hal/local/elf/platform.h:120:3: note: 'IREE_MEMORY_VIEW_FLAG_MAY_EXECUTE' declared here
2021-04-27T10:18:34.2313390Z   IREE_MEMORY_VIEW_FLAG_MAY_EXECUTE = 1u << 10,
2021-04-27T10:18:34.2367810Z   ^
2021-04-27T10:18:34.2417650Z 2 warnings and 1 error generated.
```

--- a/iree/hal/local/elf/platform/apple.c
+++ b/iree/hal/local/elf/platform/apple.c
@@ -29,6 +29,17 @@
 #include <sys/mman.h>
 #include <unistd.h>
 
+// MAP_JIT and related utilities are only available on MacOS 11.0+.
+#if defined(MAC_OS_VERSION_11_0) && \
+    MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_VERSION_11_0
+#define IREE_APPLE_IF_AT_LEAST_MAC_OS_11_0(expr) \
+  if (__builtin_available(macOS 11.0, *)) {      \
+    expr                                         \
+  }
+#else
+#define IREE_APPLE_IF_AT_LEAST_MAC_OS_11_0(expr)
+#endif  // MAC_OS_VERSION_11_0
+
 //==============================================================================
 // Memory subsystem information and control
 //==============================================================================
@@ -44,9 +55,21 @@ void iree_memory_query_info(iree_memory_info_t* out_info) {
   out_info->can_allocate_executable_pages = true;
 }
 
-void iree_memory_jit_context_begin() { pthread_jit_write_protect_np(0); }
+void iree_memory_jit_context_begin() {
+  IREE_APPLE_IF_AT_LEAST_MAC_OS_11_0({
+    if (pthread_jit_write_protect_supported_np()) {
+      pthread_jit_write_protect_np(0);
+    }
+  });
+}
 
-void iree_memory_jit_context_end() { pthread_jit_write_protect_np(1); }
+void iree_memory_jit_context_end() {
+  IREE_APPLE_IF_AT_LEAST_MAC_OS_11_0({
+    if (pthread_jit_write_protect_supported_np()) {
+      pthread_jit_write_protect_np(1);
+    }
+  });
+}
 
 //==============================================================================
 // Virtual address space manipulation
@@ -71,9 +94,11 @@ iree_status_t iree_memory_view_reserve(iree_memory_view_flags_t flags,
 
   int mmap_prot = PROT_NONE;
   int mmap_flags = MAP_PRIVATE | MAP_ANON | MAP_NORESERVE;
-  if (flags & IREE_MEMORY_VIEW_FLAG_EXECUTE) {
-    mmap_flags |= MAP_JIT;
-  }
+  IREE_APPLE_IF_AT_LEAST_MAC_OS_11_0({
+    if (flags & IREE_MEMORY_VIEW_FLAG_MAY_EXECUTE) {
+      mmap_flags |= MAP_JIT;
+    }
+  });
 
   iree_status_t status = iree_ok_status();
   void* base_address =

